src/util/taproot.rs

```diff
@@ -82,12 +93,777 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+    /// # Errors:
+    ///
+    /// - When the optimal huffman tree has a depth more than 128
+    /// - If the provided list of script weights is empty
+    pub fn new_huffman_tree<C, I>(
+        secp: &Secp256k1<C>,
+        internal_key: schnorr::PublicKey,
+        script_weights: I,
+    ) -> Result<Self, TaprootBuilderError>
+    where
+        I: IntoIterator<Item = (usize, Script)>,
+        C: secp256k1::Verification,
+    {
+        let mut node_weights = BinaryHeap::<(usize, NodeInfo)>::new();
+        for (p, leaf) in script_weights {
+            node_weights.push((p, NodeInfo::new_leaf_with_ver(leaf, LeafVersion::default())));
+        }
+        if node_weights.is_empty() {
+            return Err(TaprootBuilderError::IncompleteTree);
+        }
+        while node_weights.len() > 1 {
+            // Combine the last two elements and insert a new node
+            let (p1, s1) = node_weights.pop().expect("len must be atleast two");
+            let (p2, s2) = node_weights.pop().expect("len must be atleast two");
+            // Insert the sum of first two in the tree as a new node
+            node_weights.push((p1 + p2, NodeInfo::combine(s1, s2)?));
+        }
+        // Every iteration of the loop reduces the node_weights.len() by exactly 1
+        // Therefore, the loop will eventually terminate with exactly 1 element
+        assert!(node_weights.len() == 1);
+        let node = node_weights.pop().unwrap().1;
+        return Ok(Self::from_node_info(secp, internal_key, node));
+    }
+
+    /// Create a new key spend with internal key and proided merkle root.
+    /// Provide [`None`] for merkle_root if there is no script path.
+    /// # *Note*: As per BIP341
+    ///
+    /// When the merkle root is [`None`], the output key commits to an unspendable
+    /// script path instead of having no script path. This is achieved by computing
+    /// the output key point as Q = P + int(hashTapTweak(bytes(P)))G.
+    /// See also [`TaprootSpendInfo::tap_tweak`].
+    /// Refer to BIP 341 footnote (Why should the output key always have
+    /// a taproot commitment, even if there is no script path?) for more details
+    ///
+    pub fn new_key_spend<C: secp256k1::Verification>(
+        secp: &Secp256k1<C>,
+        internal_key: schnorr::PublicKey,
+        merkle_root: Option<sha256::Hash>,
+    ) -> Self {
+        let tweak = compute_tap_tweak(merkle_root, internal_key);
+        let mut output_key = internal_key;
+        // # Panics:
+        //
+        // This would return Err if the merkle root hash is the negation of the secret
+        // key corresponding to the internal key.
+        // Because the tweak is derived as specified in BIP341 (hash output of a function),
+        // this is unlikely to occur (1/2^128) in real life usage, it is safe to unwrap this
+        let parity = output_key
+            .tweak_add_assign(&secp, &tweak)
+            .expect("Cryptographically unreachable code");
```

Comment:
```suggestion
            .expect("compute_tap_tweak functions is broken");
```

This has much higher probability than SHA-256 collision or a tweak going above curve order :D

Reply:
Last I remember, I checked the libsecp256k1 API to make sure that `tweak_add_assign` does not when the output is out of order. We can discuss in length the exact security, but I think this is equivalent to 128-bit security. 

The point here is that this is unreachable. I can change it to "BIP 341 TapTweak is broken" if that is more descriptive. I feel the comment above does a good job of explaining why

Reply:
I think you misunderstood his point which was "since 1/2^128 is impossible, if it actually panicks it means the implementation is broken, so better use a message saying it's broken".

Reply:
I don't understand the argument. 2^{128} is cryptographically unreachable.